### PR TITLE
Fix cross build (was: fix manual)

### LINKED ISSUE #11819: Build failure in cross build to arm6l linux
**Describe the bug**

Build problem, link failure in `nix-store`

`nix build .#nix-ng-armv6l-unknown-linux-gnueabihf`

```
       > armv6l-unknown-linux-gnueabihf-g++  -o libnixstore.so libnixstore.so.p/binary-cache-store.cc.o libnixstore.so.p/build-result.cc.o libnixstore.so.p/build_derivation-goal.cc.o libnixstore.so.p/build_drv-output-substitution-goal.cc.o libnixstore.so.p/build_entry-points.cc.o libnixstore.so.p/build_goal.cc.o libnixstore.so.p/build_substitution-goal.cc.o libnixstore.so.p/build_worker.cc.o libnixstore.so.p/builtins_buildenv.cc.o libnixstore.so.p/builtins_fetchurl.cc.o libnixstore.so.p/builtins_unpack-channel.cc.o libnixstore.so.p/common-protocol.cc.o libnixstore.so.p/common-ssh-store-config.cc.o libnixstore.so.p/content-address.cc.o libnixstore.so.p/daemon.cc.o libnixstore.so.p/derivations.cc.o libnixstore.so.p/derived-path-map.cc.o libnixstore.so.p/derived-path.cc.o libnixstore.so.p/downstream-placeholder.cc.o libnixstore.so.p/dummy-store.cc.o libnixstore.so.p/export-import.cc.o libnixstore.so.p/filetransfer.cc.o libnixstore.so.p/gc.cc.o libnixstore.so.p/globals.cc.o libnixstore.so.p/http-binary-cache-store.cc.o libnixstore.so.p/indirect-root-store.cc.o libnixstore.so.p/keys.cc.o libnixstore.so.p/legacy-ssh-store.cc.o libnixstore.so.p/local-binary-cache-store.cc.o libnixstore.so.p/local-fs-store.cc.o libnixstore.so.p/local-overlay-store.cc.o libnixstore.so.p/local-store.cc.o libnixstore.so.p/log-store.cc.o libnixstore.so.p/machines.cc.o libnixstore.so.p/make-content-addressed.cc.o libnixstore.so.p/misc.cc.o libnixstore.so.p/names.cc.o libnixstore.so.p/nar-accessor.cc.o libnixstore.so.p/nar-info-disk-cache.cc.o libnixstore.so.p/nar-info.cc.o libnixstore.so.p/optimise-store.cc.o libnixstore.so.p/outputs-spec.cc.o libnixstore.so.p/parsed-derivations.cc.o libnixstore.so.p/path-info.cc.o libnixstore.so.p/path-references.cc.o libnixstore.so.p/path-with-outputs.cc.o libnixstore.so.p/path.cc.o libnixstore.so.p/pathlocks.cc.o libnixstore.so.p/posix-fs-canonicalise.cc.o libnixstore.so.p/profiles.cc.o libnixstore.so.p/realisation.cc.o libnixstore.so.p/remote-fs-accessor.cc.o libnixstore.so.p/remote-store.cc.o libnixstore.so.p/s3-binary-cache-store.cc.o libnixstore.so.p/serve-protocol-connection.cc.o libnixstore.so.p/serve-protocol.cc.o libnixstore.so.p/sqlite.cc.o libnixstore.so.p/ssh-store.cc.o libnixstore.so.p/ssh.cc.o libnixstore.so.p/store-api.cc.o libnixstore.so.p/store-reference.cc.o libnixstore.so.p/uds-remote-store.cc.o libnixstore.so.p/worker-protocol-connection.cc.o libnixstore.so.p/worker-protocol.cc.o libnixstore.so.p/linux_personality.cc.o libnixstore.so.p/unix_build_child.cc.o libnixstore.so.p/unix_build_hook-instance.cc.o libnixstore.so.p/unix_build_local-derivation-goal.cc.o libnixstore.so.p/unix_pathlocks.cc.o libnixstore.so.p/unix_user-lock.cc.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,libnixstore.so -fuse-ld=gold -Wl,--start-group /nix/store/3xc6gmhgllslbw2lhpj50ppysrm3lz1q-nix-util-armv6l-unknown-linux-gnueabihf-2.25.0pre/lib/libnixutil.so /nix/store/gan02jfaki8k9fzjbyzsdvvnlwy8kq1w-libarchive-armv6l-unknown-linux-gnueabihf-3.7.4-lib/lib/libarchive.so -pthread /nix/store/2a6ccf6r92va8a2xygx04n80228x1d8x-curl-armv6l-unknown-linux-gnueabihf-8.7.1/lib/libcurl.so /nix/store/2xhcvh6lvyr1n4251x7zf9p1grvbdj0r-libseccomp-armv6l-unknown-linux-gnueabihf-2.5.5-lib/lib/libseccomp.so /nix/store/xzn4i9mcpx6680z1c5vrslckgcpn70s2-sqlite-armv6l-unknown-linux-gnueabihf-3.45.3/lib/libsqlite3.so /nix/store/nj5mgi7vv512375q9kzmx8dycs8n1k2c-boost-armv6l-unknown-linux-gnueabihf-1.81.0/lib/libboost_container.so.1.81.0 -Wl,--end-group
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:635: error: undefined reference to '__atomic_fetch_add_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:418: error: undefined reference to '__atomic_fetch_add_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:635: error: undefined reference to '__atomic_fetch_add_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:635: error: undefined reference to '__atomic_fetch_add_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:481: error: undefined reference to '__atomic_store_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:505: error: undefined reference to '__atomic_load_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:505: error: undefined reference to '__atomic_load_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:505: error: undefined reference to '__atomic_load_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:426: error: undefined reference to '__atomic_fetch_sub_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:426: error: undefined reference to '__atomic_fetch_sub_8'
       > /nix/store/ihjldc2k25qpykr7d7jpaf7pg879q7jl-armv6l-unknown-linux-gnueabihf-gcc-13.2.0/include/c++/13.2.0/bits/atomic_base.h:426: error: undefined reference to '__atomic_fetch_sub_8'
```

**Additional context**

- Post-meson regression of https://github.com/NixOS/nix/issues/3113

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

Fix the build!

I don't think we had a manual in the cross build in the past, so this is nice.

Tested with:

```
nix build .#nix-everything-armv6l-unknown-linux-gnueabihf
```

# Context

- Closes #11819
- #2503 

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Review by Ericson2314 — APPROVED** [review]

**Comment by Mic92** [comment]
@mergify queue

**Comment by Mic92** [comment]
@mergify rebase

**Comment by edolstra** [comment]
This doesn't exactly make our Nix expressions easier to understand (and they're already massively more complex than a couple of releases ago). For instance, I have no idea what "splices" are (the Nixpkgs manual is unhelpful in this respect), so the use of functions like `makeScopeWithSplicing'`, `generateSplicesForMkScope`, etc. is a mystery to me.

**Comment by Mic92** [comment]
@roberth can you split of the first two commits into it's own PR? They are independent from @edolstra concern and they should be fixed regardless of cross-compilation or not.

File: flake.nix [hunk 1]
```diff
@@ -124,18 +124,35 @@
...
+              };
+            };
 
           # The dependencies are in their own scope, so that they don't have to be
           # in Nixpkgs top level `pkgs` or `nixComponents`.
-          nixDependencies = lib.makeScope final.newScope (import ./packaging/dependencies.nix {
-            inherit inputs stdenv;
-            pkgs = final;
-          });
+          nixDependencies = lib.makeScopeWithSplicing'
```
Comment (Mic92) [inline flake.nix:144]: So the meta-critique here is that these functions are not documented?
So let's just do that in nixpkgs?

File: flake.nix [hunk 2]
```diff
@@ -124,18 +124,35 @@
...
+          {
+            inherit (final) splicePackages;
+            inherit (final) newScope;
+          }
+          {
+            otherSplices = final.generateSplicesForMkScope "nixDependencies";
```
Comment (Mic92) [inline flake.nix:150]: This one as well.

File: flake.nix (hunk 1 above)
Comment (roberth) [reply flake.nix:144]: - https://github.com/NixOS/nixpkgs/issues/354198

**Comment by roberth** [comment]
> split

- https://github.com/NixOS/nix/pull/11825

**Comment by roberth** [comment]
> a mystery

I agree. https://github.com/NixOS/nixpkgs/issues/354198 is definitely a problem.
I might be able to simplify this with a different helper function, but the docs are the source of this trouble.

**Comment by Mic92** [comment]
@mergify rebase

**Comment by Mic92** [comment]
@roberth it would be great if I could use mergify for rebasing your pull requests, but that's not possible for the hercules-ci nix fork. Would you mind in future to just push directly to the nix repository? This avoids some back and forth.

**Comment by Ericson2314** [comment]
I would like go merge this. We shouldn't be blocked by Nixpkgs doc issues.

**Comment by roberth** [comment]
@edolstra Something like this is required to fix the cross build of `everything.nix`. The alternative would be to disable the manual, and possible run into another cross issue at a later date.

IMO the readability is a problem that should be fixed in Nixpkgs, and Nix should just follow its lead.

**Comment by roberth** [comment]
Thanks!

**Pull request merged**